interfaces.go

```diff
@@ -1,46 +1,96 @@
+	CompletionDependency DependencyType = "completion"
+	// ObviatedDependency is for dependencies that no longer have an effect due to a prior resolution.
+	// For example, if one OR is resolved, all other OR dependencies are changed to ObviatedDependency.
+	ObviatedDependency DependencyType = "obviated-dependency"
+	// SoftDependency means the dependency does not wait for this dependency to resolve.
+	// The dependency may be unresolved at the time of resolution.
+	//SoftDependency DependencyType = "soft"
+)
+
+// ResolutionStatus indicates the status of the node for situations
+//
+// The way the resolution system works is first the DAG is created, with
+// all nodes created, and all of their dependencies set.
+type ResolutionStatus string
+
+const (
+	Waiting      ResolutionStatus = "waiting"
+	Resolved     ResolutionStatus = "resolved"
+	Unresolvable ResolutionStatus = "unresolvable"
+)
+
 // DirectedGraph is the representation of a Directed Graph width nodes and directed connections.
 type DirectedGraph[NodeType any] interface {
-    // AddNode adds a node with the specified ID. If the node already exists, it returns an ErrNodeAlreadyExists.
-    AddNode(id string, item NodeType) (Node[NodeType], error)
-    // GetNodeByID returns a node with the specified ID. If the specified node does not exist, an ErrNodeNotFound is
-    // returned.
-    GetNodeByID(id string) (Node[NodeType], error)
-    // ListNodes lists all nodes in the graph.
-    ListNodes() map[string]Node[NodeType]
-    // ListNodesWithoutInboundConnections lists all nodes that do not have an inbound connection. This is useful for
-    // performing a topological sort.
-    ListNodesWithoutInboundConnections() map[string]Node[NodeType]
-    // Clone creates an independent copy of the current directed graph.
-    Clone() DirectedGraph[NodeType]
-    // HasCycles performs cycle detection and returns true if the DirectedGraph has cycles.
-    HasCycles() bool
+	// AddNode adds a node with the specified ID. If the node already exists, it returns an ErrNodeAlreadyExists.
+	AddNode(id string, item NodeType) (Node[NodeType], error)
+	// GetNodeByID returns a node with the specified ID. If the specified node does not exist, an ErrNodeNotFound is
+	// returned.
+	GetNodeByID(id string) (Node[NodeType], error)
+	// ListNodes lists all nodes in the graph.
+	ListNodes() map[string]Node[NodeType]
+	// ListNodesWithoutInboundConnections lists all nodes that do not have an inbound connection. This is useful for
+	// performing a topological sort.
+	ListNodesWithoutInboundConnections() map[string]Node[NodeType]
+	// Clone creates an independent copy of the current directed graph.
+	Clone() DirectedGraph[NodeType]
+	// HasCycles performs cycle detection and returns true if the DirectedGraph has cycles.
+	HasCycles() bool
+	// PopReadyNodes lists all nodes that have finalized their status, whether resolved
+	// or unresolvable, and clears the list.
+	PopReadyNodes() []*node[NodeType]
```

Comment:
"Lists" is an awkward verb in this context.  "Returns a list" would probably be better.

Reply:
Updated.